### PR TITLE
Scan wallet from genesis when reindexing

### LINKED ISSUE #5962: Block wallet RPC methods during reindexing
At present, it's possible to make calls to e.g. `zcash-cli getbalance` while the node is in the middle of a reindex operation. A balance obtained under these conditions is essentially certain to be invalid. We should simply block wallet-oriented RPC operations until both the reindex and the associated rescan are complete.

### LINKED ISSUE #5974: Wallet requires `-rescan` after `-reindex`
After v5.0.0, some users have reported that the wallet was left in an unusable state after starting zcashd with `-reindex`, and required a separate `-rescan` invocation to fix the wallet state. The handling of `-reindex` should ensure that `-rescan` starts correctly once reindexing is complete.

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #5962
Fixes #5974

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -6338,7 +6338,7 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
     strUsage += HelpMessageOpt("-paytxfee=<amt>", strprintf(_("Fee (in %s/kB) to add to transactions you send (default: %s)"),
                                                             CURRENCY_UNIT, FormatMoney(payTxFee.GetFeePerK())));
     strUsage += HelpMessageOpt("-rescan", _("Rescan the block chain for missing wallet transactions on startup"));
-    strUsage += HelpMessageOpt("-salvagewallet", _("Attempt to recover private keys from a corrupt wallet on startup (implies -rescan)"));
+    strUsage += HelpMessageOpt("-salvagewallet", _("Attempt to recover private keys from a corrupt wallet on startup"));
```
Comment (daira) [inline src/wallet/wallet.cpp:6345]: This still implies `-rescan`.

File: src/init.cpp [hunk 2]
```diff
@@ -965,17 +957,6 @@ void InitParameterInteraction()
             LogPrintf("%s: parameter interaction: -zapwallettxes=<mode> -> setting -rescan=1\n", __func__);
     }
 
-    if (GetBoolArg("-reindex", false) && !GetArg("-prune", 0)) {
-        if (SoftSetBoolArg("-rescan", true))
-            LogPrintf("%s: parameter interaction: -reindex=1 and not pruning -> setting -rescan=1\n", __func__);
-    }
```
Comment (str4d) [inline src/init.cpp:971]: We do want to retain something that this soft-set was doing as a side-effect: clearing the wallet's witness cache (because otherwise as the block index and chain state are rebuilt, and the chain data is replayed to the wallet, the witnesses will get out-of-sync. We should do this by passing `clearWitnessCache || fReindex` into `InitLoadWallet` below.

The soft-set here was however misleading (in that it implies that `-reindex -rescan=0` could be useful, when in fact it was a config that would break the wallet more by not clearing the witness cache), and should indeed be reverted.

File: src/init.cpp [hunk 3]
```diff
@@ -364,13 +364,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-prune=<n>", strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode disables wallet support and is incompatible with -txindex. "
             "Warning: Reverting this setting requires re-downloading the entire blockchain. "
             "(default: 0 = disable pruning blocks, >%u = target size in MiB to use for block files)"), MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024));
-#ifdef ENABLE_WALLET
-    strUsage += HelpMessageOpt("-reindex-chainstate", _("Rebuild chain state from the currently indexed blocks (implies -rescan unless pruning or unless -rescan=0 is explicitly specified)"));
-    strUsage += HelpMessageOpt("-reindex", _("Rebuild chain state and block index from the blk*.dat files on disk (implies -rescan unless pruning or unless -rescan=0 is explicitly specified)"));
-#else
```
Comment (str4d) [inline src/init.cpp:370]: After talking with @daira and staring at the documentation of `-rescan`, I think it's reasonable for this documentation to remain (i.e. say that `-reindex` implies `-rescan` for the purpose of indicating to users that a `-rescan` won't help after a `-reindex`, but not soft-setting because that's misleading). This is because `-rescan` is documented as finding missing wallet transactions, and this will naturally occur as the chain data is replayed to the wallet during a `-reindex`.

**Review by str4d — CHANGES_REQUESTED** [review]

File: src/init.cpp (hunk 2 above)
Comment (str4d) [reply src/init.cpp:971]: I tried doing this in a pairing with @therealyingtong, and then realised that `CWallet::ClearNoteWitnessCache` is currently insufficient: the Sprout and Sapling witness caches should probably be cleared (though outside of a live reorg edge case I don't think it affects much), but the Orchard tree is not being cleared, and that is I think a separate bug in the current implementation. I'm going to do some local testing, and probably open a separate issue for this (so it won't block this PR).

File: src/validationinterface.cpp [hunk 4]
```diff
@@ -170,7 +170,7 @@ void ThreadNotifyWallets(CBlockIndex *pindexLastTip)
                 blockStack.emplace_back(
                     pindex,
                     oldFrontiers,
-                    recentlyConflicted.first.at(pindex));
+                    recentlyConflicted.first[pindex]);
```
Comment (daira) [inline src/validationinterface.cpp:173]: What's the reason for this change?

(`.at` checks whether the key exists in the collection and raises an exception if not; `operator[]` has undefined behaviour in that case.)

Comment (str4d) [reply src/validationinterface.cpp:173]: `operator[]` does not have undefined behavior on this type (`std::map`); it [inserts the default value if the key doesn't exist](https://en.cppreference.com/w/cpp/container/map/operator_at). Here this would be an empty `std::vector` which has the semantics we want (no recently-conflicted transactions).
